File: cmd/app/types/api/v1/article.go
package v1

// ArticleListRequest 文章列表查询（公开）。
type ArticleListRequest struct {
	Page     int    `form:"page" binding:"omitempty,min=1"`
	PageSize int    `form:"pageSize" binding:"omitempty,min=1,max=100"`
	Keyword  string `form:"keyword" binding:"omitempty,max=200"`
	Category string `form:"category" binding:"omitempty,max=50"`
	Tag      string `form:"tag" binding:"omitempty,max=500"`
	Author   string `form:"author" binding:"omitempty,max=50"`
	Sort     string `form:"sort" binding:"omitempty,oneof=newest oldest popular"`
}

// ArticleAuthorItem 作者摘要。
type ArticleAuthorItem struct {
	ID       int64  `json:"id"`
	Username string `json:"username"`
	Nickname string `json:"nickname"`
	Avatar   string `json:"avatar"`
}

// ArticleCategoryItem 分类摘要。
type ArticleCategoryItem struct {
	ID   int64  `json:"id"`
	Name string `json:"name"`
	Slug string `json:"slug"`
}

// ArticleTagItem 标签摘要。
type ArticleTagItem struct {
	ID    int64  `json:"id"`
	Name  string `json:"name"`
	Slug  string `json:"slug"`
	Color string `json:"color"`
}

// ArticleNavItem 上一篇/下一篇。
type ArticleNavItem struct {
	ID    int64  `json:"id"`
	Title string `json:"title"`
	Slug  string `json:"slug"`
}

// ArticleListItem 列表项。
type ArticleListItem struct {
	ID           int64                `json:"id"`
	Title        string               `json:"title"`
	Slug         string               `json:"slug"`
	Summary      string               `json:"summary"`
	CoverImage   string               `json:"coverImage"`
	IsPinned     bool                 `json:"isPinned"`
	PinOrder     int                  `json:"pinOrder"`
	ViewCount    int64                `json:"viewCount"`
	LikeCount    int64                `json:"likeCount"`
	CommentCount int64                `json:"commentCount"`
	PublishedAt  string               `json:"publishedAt,omitempty"`
	UpdatedAt    string               `json:"updatedAt"`
	Author       ArticleAuthorItem    `json:"author"`
	Category     *ArticleCategoryItem `json:"category,omitempty"`
	Tags         []ArticleTagItem     `json:"tags"`
}

// ArticleListResponse 分页列表。
type ArticleListResponse struct {
	List     []ArticleListItem `json:"list"`
	Total    int64             `json:"total"`
	Page     int               `json:"page"`
	PageSize int               `json:"pageSize"`
}

// AdminArticleListRequest 管理员文章列表查询（含草稿等）。
type AdminArticleListRequest struct {
	Page     int    `form:"page" binding:"omitempty,min=1"`
	PageSize int    `form:"pageSize" binding:"omitempty,min=1,max=100"`
	Keyword  string `form:"keyword" binding:"omitempty,max=200"`
	Category string `form:"category" binding:"omitempty,max=50"`
	Tag      string `form:"tag" binding:"omitempty,max=500"`
	Author   string `form:"author" binding:"omitempty,max=50"`
	Status   string `form:"status" binding:"omitempty,max=200"` // comma-separated: draft,published,...
	Sort     string `form:"sort" binding:"omitempty,oneof=newest oldest popular"`
}

// AdminArticleListItem 管理员列表项（含状态）。
type AdminArticleListItem struct {
	ArticleListItem
	Status string `json:"status"`
}

// AdminArticleListResponse 管理员分页列表。
type AdminArticleListResponse struct {
	List     []AdminArticleListItem `json:"list"`
	Total    int64                  `json:"total"`
	Page     int                    `json:"page"`
	PageSize int                    `json:"pageSize"`
}

// ArticleDetailResponse 详情。
type ArticleDetailResponse struct {
	ArticleListItem
	Content   string          `json:"content"`
	Protected bool            `json:"protected"`
	Previous  *ArticleNavItem `json:"previous,omitempty"`
	Next      *ArticleNavItem `json:"next,omitempty"`
}

// RecordArticleViewResponse 浏览记录响应。
type RecordArticleViewResponse struct {
	ViewCount int64 `json:"viewCount"`
}

// CreateArticleRequest 创建文章（管理员）。
type CreateArticleRequest struct {
	Title       string  `json:"title" binding:"required,min=1,max=200"`
	Slug        string  `json:"slug" binding:"omitempty,max=100"`
	Content     string  `json:"content" binding:"required,min=1"`
	Summary     string  `json:"summary" binding:"omitempty,max=500"`
	CoverImage  string  `json:"coverImage" binding:"omitempty,max=500"`
	CategoryID  *int64  `json:"categoryId"`
	TagIDs      []int64 `json:"tagIds"`
	Status      string  `json:"status" binding:"omitempty,oneof=draft published archived private scheduled"`
	PublishedAt *string `json:"publishedAt" binding:"omitempty"` // RFC3339
}

// UpdateArticleRequest 更新文章（管理员，字段均可选）。
type UpdateArticleRequest struct {
	Title       *string `json:"title" binding:"omitempty,min=1,max=200"`
	Slug        *string `json:"slug" binding:"omitempty,max=100"`
	Content     *string `json:"content" binding:"omitempty,min=1"`
	Summary     *string `json:"summary" binding:"omitempty,max=500"`
	CoverImage  *string `json:"coverImage" binding:"omitempty,max=500"`
	CategoryID  *int64  `json:"categoryId"`
	TagIDs      []int64 `json:"tagIds"`
	Status      *string `json:"status" binding:"omitempty,oneof=draft published archived private scheduled"`
	PublishedAt *string `json:"publishedAt" binding:"omitempty"`
	// AutoSave 为 true 时，内容变更写入「自动保存」单槽（覆盖同文唯一一条），不递增手动版本号。
	AutoSave *bool `json:"autoSave"`
}

// UpdateArticleStatusRequest 状态变更。
type UpdateArticleStatusRequest struct {
	Status      string  `json:"status" binding:"required,oneof=draft published archived private scheduled"`
	PublishedAt *string `json:"publishedAt" binding:"omitempty"`
}

// UpdateArticleSlugRequest 别名。
type UpdateArticleSlugRequest struct {
	Slug string `json:"slug" binding:"required,min=1,max=100"`
}

// UpdateArticlePasswordRequest 密码（空字符串表示取消保护）。
type UpdateArticlePasswordRequest struct {
	Password string `json:"password" binding:"max=100"`
}

// UpdateArticlePinRequest 置顶。
type UpdateArticlePinRequest struct {
	IsPinned bool `json:"isPinned"`
	PinOrder int  `json:"pinOrder" binding:"min=0,max=1000000"`
}

// DeleteArticleResponse 删除结果。
type DeleteArticleResponse struct {
	ID int64 `json:"id"`
}

// ArticleSecurityResponse 密码更新结果（不落密文）。
type ArticleSecurityResponse struct {
	Protected bool `json:"protected"`
}

// ArticleVersionItem 单个版本摘要。
type ArticleVersionItem struct {
	ID         int64  `json:"id"`
	ArticleID  int64  `json:"articleId"`
	Title      string `json:"title"`
	Version    int    `json:"version"`
	IsAutosave bool   `json:"isAutosave"`
	CreatedBy  int64  `json:"createdBy"`
	CreatedAt  string `json:"createdAt"`
}

// ArticleVersionListResponse 版本列表。
type ArticleVersionListResponse struct {
	Items []ArticleVersionItem `json:"items"`
	Total int                  `json:"total"`
}

// UpdateArticleVersionRequest 修改版本快照显示名称（仅更新 article_versions.title）。
type UpdateArticleVersionRequest struct {
	Title string `json:"title" binding:"required,min=1,max=200"`
}

// DeleteArticleVersionResponse 删除版本记录结果。
type DeleteArticleVersionResponse struct {
	ID int64 `json:"id"`
}

// BatchArticlePayload 批量操作的附加参数。
type BatchArticlePayload struct {
	Status     string  `json:"status"`
	CategoryID *int64  `json:"categoryId"`
	TagIDs     []int64 `json:"tagIds"`
}

// BatchArticleRequest 批量操作请求。
type BatchArticleRequest struct {
	IDs     []int64             `json:"ids" binding:"required,min=1"`
	Action  string              `json:"action" binding:"required,oneof=delete set_status set_category set_tags"`
	Payload BatchArticlePayload `json:"payload"`
}

// BatchArticleResponse 批量操作结果。
type BatchArticleResponse struct {
	Affected int64 `json:"affected"`
}

File: ui/admin/src/api/article.ts
import request from '../utils/request';

export interface ArticleListQuery {
  page?: number;
  pageSize?: number;
  keyword?: string;
  category?: string;
  tag?: string;
  status?: string;
  author?: string;
  sort?: string;
}

export interface ArticleAuthorItem {
  id: number;
  username: string;
  nickname: string;
  avatar: string;
}

export interface ArticleCategoryItem {
  id: number;
  name: string;
  slug: string;
}

export interface ArticleTagItem {
  id: number;
  name: string;
  slug: string;
  color: string;
}

export interface ArticleListItem {
  id: number;
  title: string;
  slug: string;
  summary: string;
  coverImage: string;
  isPinned: boolean;
  pinOrder: number;
  viewCount: number;
  likeCount: number;
  commentCount: number;
  publishedAt: string;
  updatedAt: string;
  author: ArticleAuthorItem;
  category?: ArticleCategoryItem;
  tags: ArticleTagItem[];
}

export interface AdminArticleListItem extends ArticleListItem {
  status: string;
}

export interface ArticleDetailResponse extends AdminArticleListItem {
  content: string;
  protected: boolean;
}

/** 文章历史版本条目（管理员） */
export interface ArticleVersionItem {
  id: number;
  articleId: number;
  title: string;
  version: number;
  /** 自动保存单槽快照（同文至多一条，服务端覆盖更新） */
  isAutosave?: boolean;
  createdBy: number;
  createdAt: string;
}

export interface ArticleVersionListResponse {
  items: ArticleVersionItem[];
  total: number;
}

export interface AdminArticleListResponse {
  list: AdminArticleListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface CreateArticleRequest {
  title: string;
  slug?: string;
  content: string;
  summary?: string;
  coverImage?: string;
  categoryId?: number | null;
  tagIds?: number[];
  status?: string;
  publishedAt?: string;
}

export interface UpdateArticleRequest {
  title?: string;
  slug?: string;
  content?: string;
  summary?: string;
  coverImage?: string;
  categoryId?: number | null;
  tagIds?: number[];
  status?: string;
  publishedAt?: string;
  /** 为 true 时标题/正文变更写入自动保存单槽，不递增手动版本号 */
  autoSave?: boolean;
}

export interface BatchArticleRequest {
  action: 'delete' | 'set_status' | 'set_category' | 'set_tags';
  ids: number[];
  payload?: {
    status?: string;
    /** 传 null 表示清除分类 */
    categoryId?: number | null;
    tagIds?: number[];
  };
}

type ApiResponse<T> = { code: number; message: string; data: T };

export const getArticles = (params: ArticleListQuery) =>
  request.get<any, ApiResponse<AdminArticleListResponse>>('/api/v1/admin/articles', { params });

export const getArticle = (id: number) =>
  request.get<any, ApiResponse<ArticleDetailResponse>>(`/api/v1/admin/articles/${id}`);

export const createArticle = (data: CreateArticleRequest) =>
  request.post<any, ApiResponse<ArticleDetailResponse>>('/api/v1/admin/articles', data);

export const updateArticle = (id: number, data: UpdateArticleRequest) =>
  request.put<any, ApiResponse<ArticleDetailResponse>>(`/api/v1/admin/articles/${id}`, data);

export const listArticleVersions = (articleId: number) =>
  request.get<any, ApiResponse<ArticleVersionListResponse>>(
    `/api/v1/admin/articles/${articleId}/versions`,
  );

/** 将正文恢复到指定历史版本（直接写回，恢复前不额外生成新版本快照） */
export const restoreArticleVersion = (articleId: number, versionId: number) =>
  request.post<any, ApiResponse<ArticleDetailResponse>>(
    `/api/v1/admin/articles/${articleId}/versions/${versionId}/restore`,
  );

/** 修改版本快照的显示名称（仅更新 article_versions.title） */
export const updateArticleVersionTitle = (
  articleId: number,
  versionId: number,
  data: { title: string },
) =>
  request.patch<any, ApiResponse<ArticleVersionItem>>(
    `/api/v1/admin/articles/${articleId}/versions/${versionId}`,
    data,
  );

/** 硬删除一条版本记录 */
export const deleteArticleVersion = (articleId: number, versionId: number) =>
  request.delete<any, ApiResponse<{ id: number }>>(
    `/api/v1/admin/articles/${articleId}/versions/${versionId}`,
  );

export const deleteArticle = (id: number) =>
  request.delete<any, ApiResponse<{ id: number }>>(`/api/v1/admin/articles/${id}`);

export const batchOperateArticles = (data: BatchArticleRequest) =>
  request.post<any, ApiResponse<{ affected: number }>>('/api/v1/admin/articles/batch', data);

/** 回收站：仅已软删文章，参数与管理员列表一致（keyword、sort、分页） */
export const getTrashedArticles = (params: ArticleListQuery) =>
  request.get<any, ApiResponse<AdminArticleListResponse>>('/api/v1/admin/articles/trash', { params });

export const restoreArticle = (id: number) =>
  request.post<any, ApiResponse<{ id: number }>>(`/api/v1/admin/articles/${id}/restore`);

export const permanentDeleteArticle = (id: number) =>
  request.delete<any, ApiResponse<{ id: number }>>(`/api/v1/admin/articles/${id}/permanent`);
